### PR TITLE
cannon: Refactor and add test utils

### LINKED ISSUE #11634: cannon: Build some test tools to make testing more robust
## Description

There are a few gaps in the current Cannon test coverage:
- Many parts of the state are initialized to 0 values, so we may not be able to detect or assert on changes where we expect a value to be set to zero
- There are a lot of fields in our state structs and we are not consistently validating all of the expected values.

Add some test utilities to:
- Randomize test states so we are not always working with zero'd out fields
- Add tools to help with asserting / setting expectations for all of the fields on our state structs

### PR DESCRIPTION
**Pull request by mbaxter**
**Description**
This PR does a little bit of refactoring / cleanup and adds some new test utilities to make Cannon testing more robust and consistent.  

New utilities:
- Add an `ExpectedState` struct that:
  - Holds a set of expected values used to assert on a post-state
  - Is initialized with a pre-state, so that we need to explicitly modify any fields we expect to change
  - Provides a `Validate` method that will compare expectations with a provided post-state
- Add randomization utility:
  - Add a `WithRandomization(seed)` modifier that will randomize all fields on a state

Refactor:
- Move some utilities from `helpers.go` to `singlethreaded`, `multithreaded`, and `mipsevm` `testutil` packages to keep related code closer together / more reusable
- Move `HintTrackingOracle` logic to `testutil/oracle.go` for easier reuse

Remaining TODOs:
- https://github.com/ethereum-optimism/optimism/issues/11651
  - Created a follow-up ticket to address some remaining TODOs
  - Basically, implementing the new test utils revealed some tests that are not fully validating the post-state.  So need to do a bit more work on these tests to ensure we are fully validating expectations. 

**Metadata**

- Fixes https://github.com/ethereum-optimism/optimism/issues/11634

### CODE REVIEW

File: cannon/mipsevm/tests/helpers.go [hunk 1]
```diff
@@ -3,155 +3,31 @@ package tests
...
-	"github.com/ethereum/go-ethereum/common"
-	"github.com/ethereum/go-ethereum/common/hexutil"
 	"github.com/ethereum/go-ethereum/log"
 	"github.com/stretchr/testify/require"
 
 	"github.com/ethereum-optimism/optimism/cannon/mipsevm"
 	"github.com/ethereum-optimism/optimism/cannon/mipsevm/multithreaded"
+	testutil3 "github.com/ethereum-optimism/optimism/cannon/mipsevm/multithreaded/testutil"
 	"github.com/ethereum-optimism/optimism/cannon/mipsevm/singlethreaded"
+	testutil2 "github.com/ethereum-optimism/optimism/cannon/mipsevm/singlethreaded/testutil"
```
Comment (Inphi) [inline cannon/mipsevm/tests/helpers.go]: nit for clearer aliases:
```suggestion
	mttestutil "github.com/ethereum-optimism/optimism/cannon/mipsevm/multithreaded/testutil"
	"github.com/ethereum-optimism/optimism/cannon/mipsevm/singlethreaded"
	sttestutil "github.com/ethereum-optimism/optimism/cannon/mipsevm/singlethreaded/testutil"
```

File: cannon/mipsevm/tests/fuzz_evm_common_test.go [hunk 2]
```diff
@@ -296,57 +242,55 @@ func FuzzStateHintRead(f *testing.F) {
...
-					// Memory may be unchanged if we're writing the first zero-valued 7 bytes of the pre-image.
-					//require.NotEqual(t, preStateRoot, state.GetMemory().MerkleRoot())
-					require.Greater(t, state.GetPreimageOffset(), preimageOffset)
-				} else {
-					require.Equal(t, preStateRoot, state.GetMemory().MerkleRoot())
-					require.Equal(t, state.GetPreimageOffset(), preimageOffset)
+				// TODO(cp-983) - Do stricter validation of expected memory
+				expected.Validate(t, state, testutil.SkipMemoryValidation)
+				if writeLen == 0 {
+					// With a non-zero writeLen, we may not modify memory because the length prefix is just writing zeroes
```
Comment (Inphi) [inline cannon/mipsevm/tests/fuzz_evm_common_test.go]: nit:
```suggestion
					// With a zero writeLen, we may not modify memory because the length prefix is just writing zeroes
```

**Review by Inphi — APPROVED** [review]
